library/std/src/ffi/os_str.rs

```diff
@@ -265,6 +266,43 @@ impl OsString {
         self.inner.reserve(additional)
     }
 
+    /// Tries to reserve capacity for at least `additional` more elements to be inserted
+    /// in the given `OsString`. The collection may reserve more space to avoid
+    /// frequent reallocations. After calling `try_reserve`, capacity will be
+    /// greater than or equal to `self.len() + additional`. Does nothing if
+    /// capacity is already sufficient.
+    ///
+    /// # Errors
+    ///
+    /// If the capacity overflows, or the allocator reports a failure, then an error
+    /// is returned.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(try_reserve_2)]
+    /// use std::ffi::OsString;
+    /// use std::collections::TryReserveError;
+    ///
+    /// fn find_max_slow(data: &str) -> Result<OsString, TryReserveError> {
+    ///     let mut s = OsString::new();
+    ///
+    ///     // Pre-reserve the memory, exiting if we can't
+    ///     s.try_reserve(data.len())?;
+    ///
+    ///     // Now we know this can't OOM in the middle of our complex work
+    ///     s.push(data);
```

Comment:
I think as currently written, this would not be guaranteed not to reallocate at the push.

See the documentation of OsString::with_capacity: https://doc.rust-lang.org/1.57.0/std/ffi/struct.OsString.html#method.with_capacity &mdash; the capacity (which has the same units as `additional` here) is measured in "length units of other OS strings".

OsStr::len explains more about what "length units of an OS string" means: https://doc.rust-lang.org/1.57.0/std/ffi/struct.OsStr.html#method.len.

> The length returned is that of the underlying storage used by `OsStr`. As discussed in the `OsString` introduction, `OsString` and `OsStr` store strings in a **form best suited for cheap inter-conversion between native-platform and Rust string forms, which may differ significantly from both of them, including in storage size and encoding**.

This is saying that the **"storage size"** of an OS string (which is the size that `len` returns) **"may differ significantly"** from the **"Rust string form"** of the same data, i.e. that `os_str.len()` and `os_str.to_str().unwrap().len()` may differ significantly.

So I think try_reserve would only be meaningful if the length passed is obtained from calling `len()` on an OsStr, not a Rust str.

Reply:
> So I think try_reserve would only be meaningful if the length passed is obtained from calling len() on an OsStr, not a Rust str

Hm... the docs for [`OsString`](https://doc.rust-lang.org/1.57.0/std/ffi/struct.OsString.html) are more strongly worded:

> A type that... is cheaply inter-convertible with Rust strings

> `OsString` and `OsStr` bridge this gap by simultaneously representing Rust and platform-native string values, and in particular allowing a Rust string to be converted into an “OS” string **with no cost if possible**

Which suggests a Rust `str` length *should* be the same as an `OsStr` length, no? Or am I misreading that?

Reply:
I think that's a misreading. The units returned by str::len are bytes, while the units returned by OsStr::len are "length units of an OS string", which is not necessarily bytes. There can be a no-cost conversion even when those units are not the same.

Reply:
Ah right, sorry. I see that now.